storage/src/main/java/tech/pegasys/teku/storage/Store.java

```diff
@@ -281,6 +382,58 @@ public boolean containsCheckpointState(Checkpoint checkpoint) {
     }
   }
 
+  private Optional<BeaconState> getOrGenerateBlockState(final Bytes32 blockRoot) {
```

Comment:
We wind up running this inside while holding the read lock.  While we'll still be able to perform other read functions, all block importing would stop while this runs.  So we might want to see if we can grab the pre-state and blocks we need inside a read lock, then release the lock and do the state transitions, and finally pick up the write lock to put it back into the state cache.

Actually just realised we're using concurrent collections in Store and shouldn't be - all access is gated on the readWrite lock so the non-concurrent collections are safe to use.